### DISCUSSION TITLE
How to properly add FileManager dependency

### DISCUSSION THREAD

**Discussion by kakhaberikiknadze** [pointfreeco/swift-composable-architecture#2017] [Q&A] [answered]
I'm trying to add `FileManager` as a dependency using `Dependencies` API. However, since it's not `Sendable` I'm getting warning
``` Swift
Capture of 'fileManager' with non-sendable type 'FileManager' in a `@Sendable` closure
```

I tried to do something like this but it still won't work.

``` Swift
struct FileService: Sendable {
    var createFileInDocuments: @Sendable (AppFile) throws -> URL
}

private extension FileService {
    static var `default`: Self {
        let fileManager = FileManager.default
        return .init(
            createFileInDocuments: { @Sendable in
                try fileManager.url(
                    for: .documentDirectory,
                    in: .userDomainMask,
                    appropriateFor: nil,
                    create: true
                )
                .appending(component: $0.name)
                .appendingPathExtension($0.fileExtension)
            }
        )
    }
}
```

**Comment by tgrapperon** [comment, marked as answer]
Hey @kakhaberikiknadze!
Since `FileManager.default` is a singleton, you can probably directly use (and reuse) it from your closure without creating an instance first:
```swift
private extension FileService {
    static var `default`: Self {
        return .init(
            createFileInDocuments: { @Sendable in
                try FileManager.default.url(…)
            }
        )
    }
}
```
Of course, this trick will not work in all situations. In these cases, you will need to protect the value yourself, using `LockIsolated` that comes with `Dependencies` for example. This only makes sense if you're the owner of the private instance that is protected (which means it won't protect `FileManager.default` for example).

**Reply by kakhaberikiknadze** [reply]
Yeah, makes sense.
So far I've created `FileManagerAdapter` `actor` and added it as a dependency instead of `FileManager`.
